bindings/java/src/org/sleuthkit/datamodel/BlackboardAttribute.java

```diff
@@ -483,33 +304,39 @@ public void addSource(String source) throws TskCoreException {
 	 */
 	public BlackboardArtifact getParentArtifact() throws TskCoreException {
 		if (parentArtifact == null) {
-			parentArtifact = sleuthkitCase.getBlackboardArtifact(artifactID);
+			parentArtifact = getSleuthkitCase().getBlackboardArtifact(getArtifactID());
 		}
 		return parentArtifact;
 	}
 
+
 	@Override
 	public int hashCode() {
-		int hash = 5;
-		hash = 97 * hash + (int) (this.artifactID ^ (this.artifactID >>> 32));
-		return hash;
+		return Objects.hash(
+				this.getAttributeType(), this.getValueInt(), this.getValueLong(), this.getValueDouble(),
+				this.getValueString(), this.getValueBytes(), this.getSources());
 	}
 
 	@Override
-	public boolean equals(Object obj) {
-		if (obj == null) {
-			return false;
-		}
-		if (getClass() != obj.getClass()) {
+	public boolean equals(Object that) {
+		if (this == that) {
+			return true;
+		} else if (that instanceof BlackboardAttribute) {
+			BlackboardAttribute other = (BlackboardAttribute) that;
+			Object[] thisObject = new Object[]{this.getAttributeType(), this.getValueInt(), this.getValueLong(), this.getValueDouble(),
+				this.getValueString(), this.getValueBytes(), this.getSources()};
+			Object[] otherObject = new Object[]{other.getAttributeType(), other.getValueInt(), other.getValueLong(), other.getValueDouble(),
+				other.getValueString(), other.getValueBytes(), other.getSources()};
+
+			return Objects.deepEquals(thisObject, otherObject);
+		} else {
```

Comment:
We have some duplicated code here. Perhaps we can create a protected helper in the super class for the comparison of the attribute type and values? Also, I do not see code to compare the unique fields of this class, i.e., context, sources, artifact ID, etc.